### ISSUE TITLE
/bin:/usr/bin in PATH causes failure in certain GUI apps (LinuxMint 21)

### ISSUE THREAD

**Issue by jrw** [linuxmint/cinnamon#11582]
```
 * Cinnamon version (cinnamon --version)
   - Please specify if you are using the daily builds PPA
     (https://launchpad.net/~linuxmint-daily-build-team/+archive/ubuntu/daily-builds).
 * Distribution - (Mint 21, Arch, Fedora 37, etc...)
 * Graphics hardware *and* driver used
 * 32 or 64 bit
 * Attach ~/.xsession-errors, or /var/log/syslog
 ```
Cinnamon 5.4.12
LinuxMint 21
I don't believe this is graphics or 32/64 bit related.

### Issue

Apparently, the order of `/bin` and `/usr/bin` in `PATH` matters!  I was getting failures for multiple apps started from LinuxMint's Menu:  mintsources, lightdm-settings, etc.  Eventually I tracked it down to the fact that I had `/bin:/usr/bin` in my PATH rather than `/usr/bin:/bin`.  Once I switched the order in my PATH and rebooted, the GUI apps started working.

### Steps to reproduce

$ ls -ld /bin
lrwxrwxrwx 1 root root 7 Jun 28  2022 /bin -> usr/bin

$ pkexec /usr/bin/mintsources
...GUI prompt for password, then GUI app appears and works...

$ pkexec /bin/mintsources
...GUI prompt for password, then...
(mintSources.py:3926): Gtk-CRITICAL **: 08:24:53.938: _gtk_style_provider_private_get_settings: assertion 'GTK_IS_STYLE_PROVIDER_PRIVATE (provider)' failed
(mintSources.py:3926): Gtk-CRITICAL **: 08:24:53.938: _gtk_style_provider_private_get_settings: assertion 'GTK_IS_STYLE_PROVIDER_PRIVATE (provider)' failed
(mintSources.py:3926): Gtk-CRITICAL **: 08:24:53.938: _gtk_style_provider_private_get_settings: assertion 'GTK_IS_STYLE_PROVIDER_PRIVATE (provider)' failed
...ABORTS (no GUI app appears)...

### Expected behaviour

Both should work identically and successfully.  Currently, the order of `/usr/bin:/bin` in `PATH` is required for working apps and the order of `/bin:/usr/bin` causes them to fail.  The order shouldn't matter.

### Other information

See more discussion and bug confirmation at [LinuxMint forum](https://forums.linuxmint.com/viewtopic.php?p=2310204#p2310204)

### Configuration info, if applicable
If this bug report is related to an Applet, Desklet or Extension, please paste (or use a pastebin service) the offending extension's settings, if it has any.  You can obtain this by opening its configuration, clicking the 'hamburger' button in the upper-right corner of the window, and selecting "Export to a file". **Please be sure to review the contents and remove any personal data it may contain.**

**Comment by mtwebster** [comment]
This all seems 'expected' (though annoying I suppose) -

It's really just how policykit (pkexec) works:

If you look in `/usr/share/polkit-1/actions/com.linuxmint.mintsources.policy`, at the bottom of the file you see:

```
...
    <annotate key="org.freedesktop.policykit.exec.path">/usr/bin/mintsources</annotate>
    <annotate key="org.freedesktop.policykit.exec.allow_gui">true</annotate>
...
```
If you look at the [pkexec documentation](https://www.freedesktop.org/software/polkit/docs/latest/pkexec.1.html#pkexec-action) `, under Required Authorizations, it says `... use the org.freedesktop.policykit.exec.path annotation on an action with the value set to the full path of the program....`.

If you run `pkexec /bin/mintsources`, it doesn't match the exec.path, and you end up running mintsources without a sufficient set of environment variables - including those needed for Gtk programs to run properly. It also doesn't work if you simply put 'mintsources' there (I tried to be sure).

I changed the annotate keys to point to `/usr/bin/sh` instead, then ran pkexec on it:
```
$ pkexec /usr/bin/sh
# env
LANGUAGE=en_US
LC_TIME=en_US.UTF-8
USER=root
HOME=/root
LC_MONETARY=en_US.UTF-8
COLORTERM=truecolor
LOGNAME=root
TERM=xterm-256color
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
PKEXEC_UID=1000
DISPLAY=:0
LANG=en_US.UTF-8
XAUTHORITY=/home/mtwebster/.Xauthority
SHELL=/bin/bash
PWD=/root
LC_NUMERIC=en_US.UTF-8
```
Then I tried `pkexec /bin/sh`:
```
$ pkexec /bin/sh
# env     
LANGUAGE=en_US
LC_TIME=en_US.UTF-8
USER=root
HOME=/root
LC_MONETARY=en_US.UTF-8
COLORTERM=truecolor
LOGNAME=root
TERM=xterm-256color
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
PKEXEC_UID=1000
LANG=en_US.UTF-8
SHELL=/bin/bash
PWD=/root
LC_NUMERIC=en_US.UTF-8
```

Two important things are missing from the second (incorrect) one:
```
DISPLAY=:0
XAUTHORITY=/home/mtwebster/.Xauthority
```
Without DISPLAY set, gui programs will fail.

I'm guessing this issue has only come about since the 'usrmerge' occurred - prior to that, mintsources would not be found in `/bin`, only `/usr/bin`, so your path order wouldn't matter. Now, it *is* there, and even though /bin is just a symbolic link to /usr/bin, pkexec takes the path at face value, and won't allow it (a good thing, I think).

As for the menu, the mintsources desktop file (what the menu reads from) runs `pkexec mintsources`, which in your case resolves to /bin/mintsources, explaining that failure.

Was there a reason you needed to change the path order system-wide? Would only setting it for shell sessions work instead?

**Comment by jrw** [comment]
@mtwebster Wow, thank you for that investigation and explanation.  I wasn't sure of the mechanism, but I could see there was something like that happening and I had also discovered that using `pkexec env -i PATH=$PATH DISPLAY=$DISPLAY XAUTHORITY=$XAUTHORITY mintsources` would work, even with my original PATH.  (See this link to my [linuxmint forum post](https://forums.linuxmint.com/viewtopic.php?p=2310204) which contains the full history of this bug.)

By "change the path order system-wide", I'm not sure what you mean.  I didn't change anything in `/etc/login.defs` or anything like that.  I have been customizing PATH in my `.bashrc`, `.bash_profile`, `.kshrc`, `.profile` for probably 30 years.  I fully rewrite the PATH, including the order that I have been using of `/bin:/usr/bin`.  After the usrmerge, everything continued to work as expected (except in this case!).  So I have now changed the order of my PATH components to `/usr/bin:/bin` when I rewrite PATH in my shell profile(s).  I guess that at some point in the login process that my shell profile gets executed to create the environment that launchers are run in.  In other words, it appears that the environment for launchers is controlled by my shell profile.  So, after rebooting with my new profile, things are working again.

It looks like this is a bug with pkexec and/or the polkit mechanism:  `pkexec` doesn't fully resolve (canonicalize) the path to the executable, a la `readlink -f /bin/mintsources`, before matching it against the path in the policy.  The `pkexec` documentation says `full path of the program` where I think it should say `canonical full path of the program`.

In any case, apparently this bug is not too serious since very few people are hitting it sufficiently hard to report it or track it down.  I was unable to duck-duck-go any information that would have led me to the fix, so I've been living with these broken launchers for several years (probably since the usrmerge) and only with my recent upgrade to Mint 21 did I muster up enough persistence to finally track it down.

I guess this means I should report this upstream to Ubuntu and the `pkexec` package?

P.S. I'm not sure why pkexec works for executables when it can't find a policy for that executable (because their path doesn't match the configured path of a policy).  As your test shows, it ends up with an incorrect result when it does so.
